### PR TITLE
bugfix: Fix Gatling Cannon barrels rotating despite insufficient energy

### PR DESCRIPTION
**Pull request by bobtista**
Addresses issue #1700  

Tested by building a power plant, then a gatling cannon, force firing it, then selling the power plant and trying to force fire.

Result: The gatling cannon can no longer force fire while powered down, the "can't" mouse cursor is displayed, and there is no spinning animation.

### CODE REVIEW

File: Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 1]
```diff
@@ -2881,6 +2881,10 @@ Bool Object::isAbleToAttack() const
 	if( testStatus(OBJECT_STATUS_SOLD) )
 		return false;
 
+	// TheSuperHackers @bugfix bobtista 31/10/2025 Fixes Gatling Cannon barrels rotating despite insufficient energy. (#1700)
```
Comment (Skyaero42) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp]: Remove issue number

File: Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 2]
```diff
@@ -2881,6 +2881,10 @@ Bool Object::isAbleToAttack() const
...
+	if ( isKindOf( KINDOF_POWERED ) && isDisabledByType( DISABLED_UNDERPOWERED ) )
```
Comment (Skyaero42) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp]: So previously, patriot and strat centre would also show attack mouse marker rather than disabled mouse marker?

**Review by Skyaero42 — DISMISSED** [review]

File: Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp (hunk 2 above)
Comment (bobtista) [reply Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp]: Yes - in retail, Patriots (and any other KINDOF_POWERED structures) would also show the attack mouse marker rather than the disabled mouse marker when underpowered.
However, the Strategy center cannon is not a powered thing, it can still force fire like a fire base without power. 
I've tested these with this fix - while undepowered, the patriot and gatling cannon now correctly show the disabled mouse cursor, and the strategy center can still force fire its cannon.

**Comment by Mauller** [comment]
This wants testing for retail compatibility

**Comment by Skyaero42** [comment]
> This wants testing for retail compatibility

This one’s tricky. I ran a comparison test:

- Played a vanilla game where I force-fired the Gatling.
- Then ran that replay on this PR build 
- (and vice versa — though force-fire doesn’t work here, so that test isn’t very meaningful).

When playing in vanilla and replaying in this PR, the barrels still move when the force-attack command was given. That suggests the barrel rotation state is recorded in the replay. However, there is no mismatch, so maybe it is not transferred in the xfer.

To stay safe, I’d recommend putting this change behind CRC guards.

Also, there’s an edge case still present in this PR:
- Build Power and a Gatling.
- Force-fire the Gatling.
- Sell the Power.
→ Once power is gone, the Gatling stops spinning, but starts again after about two seconds — which shouldn’t happen.

**Comment by bobtista** [comment]
> > This wants testing for retail compatibility
> 
> This one’s tricky. I ran a comparison test:
> 
> * Played a vanilla game where I force-fired the Gatling.
> * Then ran that replay on this PR build
> * (and vice versa — though force-fire doesn’t work here, so that test isn’t very meaningful).
> 
> When playing in vanilla and replaying in this PR, the barrels still move when the force-attack command was given. That suggests the barrel rotation state is recorded in the replay. However, there is no mismatch, so maybe it is not transferred in the xfer.
> 
> To stay safe, I’d recommend putting this change behind CRC guards.
> 
> Also, there’s an edge case still present in this PR:
> 
> * Build Power and a Gatling.
> * Force-fire the Gatling.
> * Sell the Power.
>   → Once power is gone, the Gatling stops spinning, but starts again after about two seconds — which shouldn’t happen.

I added another check in adjustModelConditionForWeaponStatus() that hopefully addresses the spinning animation edge case - I can't test it this week though, if someone else can try it. I put the change behind CRC guards

**Comment by Skyaero42** [comment]
> 
    Build Power and a Gatling.
    Force-fire the Gatling.
    Sell the Power.
    → Once power is gone, the Gatling stops spinning, but starts again after about two seconds — which shouldn’t happen.


This still occurs

**Comment by Skyaero42** [comment]
Have you tested it?

**Comment by bobtista** [comment]
> Have you tested it?
Yes, the issue persists. I've been working on it today, I think I might be close to a fix.

**Comment by Caball009** [comment]
This might require a slightly different approach. I think I'm seeing the correct results when I do this:
1. Keep only the new addition to `Object::isAbleToAttack`.
2. Add this to the end of `Object::setDisabledUntil`:
```cpp
//clearStatus( MAKE_OBJECT_STATUS_MASK( OBJECT_STATUS_IS_ATTACKING ) );
clearModelConditionState( MODELCONDITION_ATTACKING );
```

FWIW, for quicker or alternative testing, you can use microwave tanks to disable either a power plant or disable the gattling cannon itself.

**Comment by bobtista** [comment]
> This might require a slightly different approach. I think I'm seeing the correct results when I do this:
> 
> 1. Keep only the new addition to `Object::isAbleToAttack`.
> 2. Add this to the end of `Object::setDisabledUntil`:
> 
> ```c++
> //clearStatus( MAKE_OBJECT_STATUS_MASK( OBJECT_STATUS_IS_ATTACKING ) );
> clearModelConditionState( MODELCONDITION_ATTACKING );
> ```
> 
> FWIW, for quicker or alternative testing, you can use microwave tanks to disable either a power plant or disable the gattling cannon itself.

I spent all day yesterday troubleshooting this - with your fix, the barrels still rotate for a moment just after the power goes out. They go from "power on, firing" to "power off, stopped" for a second, to "power off, spinning" for a second, then finally they stop.

**Comment by Caball009** [comment]
> with your fix, the barrels still rotate for a moment just after the power goes out.

I'm seeing the same behavior. I tried a couple of things, but was unable to fix this.

**Comment by Caball009** [comment]
Could maybe be implemented like this: https://github.com/Caball009/GeneralsGameCode/tree/fix-gatling-barrel-animation-no-power2

The barrels spinning up shortly after the power goes out is because of `FiringTracker::coolDown`, which is called at intervals so that accounts for the short delay.

**Comment by Caball009** [comment]
Maybe a cleaner solution would be to modify `FiringTracker::getDisabledTypesToProcess`, so that firing tracker isn't updated when the parent object is disabled.

**Comment by Caball009** [comment]
(I didn't realize the game was using the incorrect spelling for 'gatling').

File: Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 3]
```diff
@@ -1558,6 +1558,16 @@ Bool Object::isLogicallyVisible() const
 //=============================================================================
 // Object::isLocallyControlled
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp:1559]: Wrong comment placement

File: Generals/Code/GameEngine/Source/GameLogic/Object/FiringTracker.cpp [hunk 4]
```diff
@@ -282,6 +282,13 @@ void FiringTracker::speedUp()
 
 }
 
+//-------------------------------------------------------------------------------------------------
+void FiringTracker::forceCoolDown()
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/FiringTracker.cpp:286]: Maybe name this coolDownNow

File: Generals/Code/GameEngine/Include/GameLogic/FiringTracker.h [hunk 5]
```diff
@@ -55,9 +55,21 @@ class FiringTracker : public UpdateModule
 	void shotFired(const Weapon* weaponFired, ObjectID victimID );			///< Owner just fired this weapon at this Object
 	ObjectID getLastShotVictim() const { return m_victimID; }						///< get the last victim ID that was shot at
 	Int getNumConsecutiveShotsAtVictim( const Object *victim ) const;
+	void forceCoolDown();																						///< Force immediate cooldown, stopping all continuous fire states
 
+#if !RETAIL_COMPATIBLE_CRC
+	/// Exclude power-related disable types so update() doesn't restart barrel animations.
+	/// forceCoolDown() in setDisabledUntil() handles immediate cooldown.
+	virtual DisabledMaskType getDisabledTypesToProcess() const override
```
Comment (xezon) [inline Generals/Code/GameEngine/Include/GameLogic/FiringTracker.h:63]: Better move implementation details to CPP to keep the header tidy.

File: Generals/Code/GameEngine/Include/GameLogic/FiringTracker.h [hunk 6]
```diff
@@ -55,9 +55,21 @@ class FiringTracker : public UpdateModule
 	void shotFired(const Weapon* weaponFired, ObjectID victimID );			///< Owner just fired this weapon at this Object
 	ObjectID getLastShotVictim() const { return m_victimID; }						///< get the last victim ID that was shot at
 	Int getNumConsecutiveShotsAtVictim( const Object *victim ) const;
+	void forceCoolDown();																						///< Force immediate cooldown, stopping all continuous fire states
 
+#if !RETAIL_COMPATIBLE_CRC
```
Comment (xezon) [inline Generals/Code/GameEngine/Include/GameLogic/FiringTracker.h:60]: Prefer using positive order `#if RETAIL_COMPATIBLE_CRC`

Several times in this change.

File: Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 7]
```diff
@@ -1558,6 +1558,16 @@ Bool Object::isLogicallyVisible() const
...
 //=============================================================================
+Bool Object::isUnderpoweredForAttack() const
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp:1561]: This function is superfluous. Just put `isDisabledByType( DISABLED_UNDERPOWERED )` at the call sites.

Or remove RETAIL_COMPATIBLE_CRC from all the callsites.

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 8]
```diff
@@ -2236,6 +2246,20 @@ void Object::setDisabledUntil( DisabledType type, UnsignedInt frame )
 
 	}
 
+#if !RETAIL_COMPATIBLE_CRC
+	// TheSuperHackers @bugfix bobtista 21/12/2025 Fix Gatling Cannon barrels rotating despite insufficient energy.
+	// When power is lost (UNDERPOWERED, EMP, SUBDUED, HACKED), immediately force FiringTracker cooldown to stop barrel animations.
+	// getDisabledTypesToProcess() prevents update() from restarting animations, and isUnderpoweredForAttack() prevents cursor/attack logic.
+	if (m_firingTracker)
+	{
+		Bool isPowerDisableType = (type == DISABLED_UNDERPOWERED || type == DISABLED_EMP || type == DISABLED_SUBDUED || type == DISABLED_HACKED);
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp:2255]: Better order these by their enum order

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 9]
```diff
@@ -2236,6 +2246,20 @@ void Object::setDisabledUntil( DisabledType type, UnsignedInt frame )
 
 	}
 
+#if !RETAIL_COMPATIBLE_CRC
+	// TheSuperHackers @bugfix bobtista 21/12/2025 Fix Gatling Cannon barrels rotating despite insufficient energy.
+	// When power is lost (UNDERPOWERED, EMP, SUBDUED, HACKED), immediately force FiringTracker cooldown to stop barrel animations.
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp:2251]: Better not list the enum values in the comment, because it is redundant information and needs updating if a flag was added or removed.

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 10]
```diff
@@ -2236,6 +2246,20 @@ void Object::setDisabledUntil( DisabledType type, UnsignedInt frame )
...
+	// getDisabledTypesToProcess() prevents update() from restarting animations, and isUnderpoweredForAttack() prevents cursor/attack logic.
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp:2252]: Note that these comments will need updating if the functions are renamed or current logic moves into other functions. Fundamentally, this kind of verbose explaining where logic lives elsewhere will lead to scenarios where outdated comments will claim something which is not actually true after the code was changed later.

File: Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 11]
```diff
@@ -1558,6 +1558,16 @@ Bool Object::isLogicallyVisible() const
...
+{
+#if !RETAIL_COMPATIBLE_CRC
+	return isKindOf( KINDOF_POWERED ) && isDisabledByType( DISABLED_UNDERPOWERED );
```
Comment (xezon) [inline Generals/Code/GameEngine/Source/GameLogic/Object/Object.cpp:1564]: Is the `KINDOF_POWERED` really necessary here? It looks redundant, because

```cpp
static void doPowerDisable( Object *obj, void *userData )
{
	Bool disabling = *((Bool*)userData);
	if( obj && obj->isKindOf(KINDOF_POWERED) ) // <----- is guarded by KINDOF_POWERED
	{
		if( disabling )
			obj->setDisabled( DISABLED_UNDERPOWERED ); //set disabled has a pauseAllSpecialPowers that prevents double pausing
		else
			obj->clearDisabled( DISABLED_UNDERPOWERED );
	}
}
```

Instead, maybe just put an assert that if DISABLED_UNDERPOWERED, then KINDOF_POWERED is expected to be set.

File: GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp [hunk 12]
```diff
@@ -3241,6 +3265,11 @@ Bool Object::isAbleToAttack() const
   if ( isDisabledByType( DISABLED_SUBDUED ) )
     return FALSE; // A Microwave Tank is cooking me
 
+#if !RETAIL_COMPATIBLE_CRC
+	if (isUnderpoweredForAttack())
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/Object/Object.cpp:3269]: What about `isDisabledByType( DISABLED_EMP )`? Wouldn't that also be required for Gatling Cannons?

Image

**Comment by Skyaero42** [comment]
Using force-attack, the turret still starts to spin when it is low power.

I feel like the author isn't testing its solutions - and with that wasting my time. I've now spend a significant amount of time testing and retesting commits in this PR and finding that whatever is supposed to be solved, isn't.

I will not longer review this PR. I'm not putting any effort in, if the author isn't either.

**Comment by bobtista** [comment]
Seems more trouble than it's worth, it's purely a visual thing
